types/prismjs/index.d.ts

```diff
@@ -1,228 +1,435 @@
-export type AvailableHooks
-    = "before-highlightall"
-    | "before-sanity-check"
-    | "before-highlight"
-    | "before-insert"
-    | "after-highlight"
-    | "complete"
-    | "wrap";
-
-export interface Hooks {
-    all: Array<Array<(env: Environment) => void>>;
-
-    add(name: AvailableHooks | string, callback: HookCallback): void;
-
-    run(name: AvailableHooks | string, env: Environment): void;
+export namespace hooks {
+	/**
+	 * @param env The environment variables of the hook.
+	 */
+	type HookCallback = (env: Environment) => void;
+	type HookTypes = keyof HookEnvironmentMap;
+
+	interface HookEnvironmentMap {
+		"before-highlightall": RequiredEnvironment<"selector">;
+
+		"before-sanity-check": ElementEnvironment;
+		"before-highlight": ElementEnvironment;
+
+		"before-insert": ElementHighlightedEnvironment;
+		"after-highlight": ElementHighlightedEnvironment;
+		complete: ElementHighlightedEnvironment;
+
+		"before-tokenize": TokenizeEnvironment;
+		"after-tokenize": TokenizeEnvironment;
+
+		wrap: RequiredEnvironment<
+			"type" | "content" | "tag" | "classes" | "attributes" | "language"
+		>;
+	}
+
+	type RequiredEnvironment<
+		T extends keyof Environment,
+		U extends Environment = Environment
+	> = U & Required<Pick<U, T>>;
+	type ElementEnvironment = RequiredEnvironment<
+		"element" | "language" | "grammar" | "code"
+	>;
+	type ElementHighlightedEnvironment = RequiredEnvironment<
+		"highlightedCode",
+		ElementEnvironment
+	>;
+	type TokenizeEnvironment = RequiredEnvironment<
+		"code" | "grammar" | "language"
+	>;
+
+	interface RegisteredHooks {
+		[hook: string]: HookCallback[];
+	}
+
+	const all: RegisteredHooks;
+
+	/**
+	 * Adds the given callback to the list of callbacks for the given hook.
+	 *
+	 * The callback will be invoked when the hook it is registered for is run.
+	 * Hooks are usually directly run by a highlight function but you can also run hooks yourself.
+	 *
+	 * One callback function can be registered to multiple hooks and the same hook multiple times.
+	 *
+	 * @param name The name of the hook.
+	 * @param callback The callback function which is given environment variables.
+	 */
+	function add<K extends keyof HookEnvironmentMap>(
+		name: K,
+		callback: (env: HookEnvironmentMap[K]) => void
+	): void;
+	function add(name: string, callback: HookCallback): void;
+
+	/**
+	 * Runs a hook invoking all registered callbacks with the given environment variables.
+	 *
+	 * Callbacks will be invoked synchronously and in the order in which they were registered.
+	 *
+	 * @param name The name of the hook.
+	 * @param env The environment variables of the hook passed to all callbacks registered.
+	 */
+	function run<K extends keyof HookEnvironmentMap>(
+		name: K,
+		env: HookEnvironmentMap[K]
+	): void;
+	function run(name: string, env: Environment): void;
 }
 
-export type TokenNode = Token | string | Array<Token | string>;
+export type TokenNode = string | Token | Array<string | Token>;
```

Comment:
Wouldn't `TokenStream` be a better name?

Reply:
Probably.

You should probably add it to `prism-core.doc.js` in https://github.com/PrismJS/prism/pull/1782.